### PR TITLE
add support for --extended-dry-run/-x (REVIEW)

### PR DESCRIPTION
**Pull request by boegel**
TODO:

* [x] add warning that actual build procedure may diverge from output shown, due to conditional checks in the easyblock
* [x] catch `EasyBuildError` for each method being called, simply report the error and continue?
* [x] add `(DRY RUN)` tag for `== ` line printed for every step
* [x] auto-enable `--force`?
* [x] sort env vars + avoid `$EBVAR*` (see also #1219)
* [x] verbose dry run of `sanity_check_step`
* [x] verbose dry run of `make_module_step`
* [x] more verbose dry run version of `fetch_step`
* [x] avoid excessive output from `extensions_step`
* [x] simulate `module load` for deps/toolchain if module is missing
* [x] update documentation (see https://github.com/hpcugent/easybuild/pull/159)

Example output:

```
$ eb make-3.82-GCC-4.8.2.eb -x
== temporary log file in case of crash /tmp/eb-ksVC07/easybuild-MmAugb.log

== processing EasyBuild easyconfig /home/example/eb/easybuild-easyconfigs/easybuild/easyconfigs/m/make/make-3.82-GCC-4.8.2.eb

Important note: the actual build & install procedure that will be performed may diverge
(slightly) from what is outlined below, due to conditions in the easyblock which are
incorrectly handled in a dry run.
Any errors that may occur are ignored and reported as warnings, on a per-step basis.
Please be aware of this, and only use the information below for quick debugging purposes.

*** DRY RUN using 'ConfigureMake' easyblock (easybuild.easyblocks.generic.configuremake @ /home/example/eb/easybuild-easyblocks/easybuild/easyblocks/generic/configuremake.pyc) ***

== building and installing make/3.82-GCC-4.8.2...
fetching files... [DRY RUN]

[fetch_step method]
Available download URLs for sources/patches:
  * http://ftpmirror.gnu.org/make/$source

List of sources:
  * make-3.82.tar.bz2 will be downloaded to /home/example/eb/sources/m/make/make-3.82.tar.bz2

List of patches:
(none)

creating build dir, resetting environment... [DRY RUN]

[check_readiness_step method]

[make_builddir method]

[reset_changes method]

[handle_iterate_opts method]

unpacking... [DRY RUN]

[checksum_step method]
* expected checksum for make-3.82.tar.bz2: (none)

[extract_step method]
  running command "tar xjf /home/example/eb/sources/m/make/make-3.82.tar.bz2"
  (in /home/example/eb/build/make/3.82/GCC-4.8.2)

patching... [DRY RUN]

[patch_step method]

preparing... [DRY RUN]

[prepare_step method]
Defining build environment, based on toolchain (options) and specified dependencies...

Loading toolchain module...

module load GCC/4.8.2

Loading modules for dependencies...


Full list of loaded modules:
  1) GCC/4.8.2

Defining build environment...

  export CC="gcc"
  export CFLAGS="-O2"
  export CXX="g++"
  export CXXFLAGS="-O2"
  export F77="gfortran"
  export F90="gfortran"
  export F90FLAGS="-O2"
  export FFLAGS="-O2"
  export FLIBS="-lgfortran"
  export LDFLAGS="-L/home/example/eb/software/GCC/4.8.2/lib"
  export LIBS="-lm -lpthread"
  export OPTFLAGS="-O2"
  export PRECFLAGS=""

configuring... [DRY RUN]

[configure_step method]
  running command " ./configure --prefix=/home/example/eb/software/make/3.82-GCC-4.8.2 "
  (in /home/example/eb/build/make/3.82/GCC-4.8.2/make-3.82)

building... [DRY RUN]

[build_step method]
  running command " make -j 4 "
  (in /home/example/eb/build/make/3.82/GCC-4.8.2/make-3.82)

testing... [DRY RUN]

[test_step method]

installing... [DRY RUN]

[stage_install_step method]

[make_installdir method]

[install_step method]
  running command " make install "
  (in /home/example/eb/build/make/3.82/GCC-4.8.2/make-3.82)

taking care of extensions... [DRY RUN]

[extensions_step method]

postprocessing... [DRY RUN]

[post_install_step method]

sanity checking... [DRY RUN]

[sanity_check_step method]
Sanity check paths - file ['files']
  * bin/make
Sanity check paths - (non-empty) directory ['dirs']
  (none)
Sanity check commands
  (none)

cleaning up... [DRY RUN]

[cleanup_step method]

creating module... [DRY RUN]

[make_module_step method]
Generating module file /home/example/eb/modules/all/make/3.82-GCC-4.8.2, with contents:

    #%Module
    proc ModulesHelp { } {
        puts stderr { make-3.82: GNU version of make utility - Homepage: http://www.gnu.org/software/make/make.html
        }
    }
    
    module-whatis {Description: make-3.82: GNU version of make utility - Homepage: http://www.gnu.org/software/make/make.html}
    
    set root /home/example/eb/software/make/3.82-GCC-4.8.2
    
    conflict make
    
    if { ![ is-loaded GCC/4.8.2 ] } {
        module load GCC/4.8.2
    }
    
    setenv	EBROOTMAKE		"$root"
    setenv	EBVERSIONMAKE		"3.82"
    setenv	EBDEVELMAKE		"$root/easybuild/make-3.82-GCC-4.8.2-easybuild-devel"
    
    # Built with EasyBuild version 2.4.0
    

permissions... [DRY RUN]

[permissions_step method]

packaging... [DRY RUN]

[package_step method]

== COMPLETED: Installation ended successfully

Important note: the actual build & install procedure that will be performed may diverge
(slightly) from what is outlined above, due to conditions in the easyblock which are
incorrectly handled in a dry run.
Any errors that may occur are ignored and reported as warnings, on a per-step basis.
Please be aware of this, and only use the information above for quick debugging purposes.

(no ignored errors during dry run)

== Build succeeded for 1 out of 1
== Temporary log file(s) /tmp/eb-ksVC07/easybuild-MmAugb.log* have been removed.
== Temporary directory /tmp/eb-ksVC07 has been removed.
```

### CODE REVIEW

**Comment by pforai** [comment]
woot! great visibility improvement. totally dig this. :+1:

**Comment by verdurin** [comment]
Looks very useful.

Occurs to me that showing where files would be fetched from would be a good addition.

**Comment by boegel** [comment]
@verdurin: you mean download URL and path where it'll be downloaded to?

**Comment by verdurin** [comment]
@boegel - exactly, yes.

**Comment by pescobar** [comment]
this looks really useful.

I would add to verdurin comment that if the source tarball is found locally it would be nice to show the full path to the source file.

**Comment by boegel** [comment]
I think I'll go with something like:

```
[fetch_step method]
List of required files:
* make-3.82.tar.bz2 found at /home/kehoste/easybuild/sources/make-3.82.tar.bz2
```

vs

```
[fetch_step method]
List of required files:
* make-3.82.tar.bz2 downloaded from http://ftpmirror.gnu.org/make/make-3.82.tar.bz2 to /home/kehoste/easybuild/sources/make-3.82.tar.bz2
```

**Comment by verdurin** [comment]
Yes, that's what's needed here.

**Comment by fgeorgatos** [comment]
@boegel: nice one, this will help a lot plenty of people, good job!

**Comment by gppezzi** [comment]
+1 for this feature!

Can we use the output from this for reproducing the build environment on the current shell? It would be nice for debugging/tweaking the parameters of a failing build.

**Comment by boegel** [comment]
@gppezzi: i can tweak the format of the output a bit so you can just copy-paste this, yeah...

E.g.:

```
export FOO="bar baz"
module load dep1/1.2.3
```

**Comment by boegel** [comment]
@pescobar, @verdurin: hmm, while looking into making the output for `fetch` more verbose as requested, I realised that things are more complicated: if multiple source URLs are listed, they're cycled over when trying to download a file... So, we don't know where files will be downloaded from until we actually try it.

I could make the download mechanism itself support a 'dry run' mode, where only the HTTP return code is checked rather than actually doing the download, but that implies that you need to be online to use `-x` if you don't have the files available yet.

So, counter proposal:

```
[fetch_step method]
Download URLs (for files not available yet):
* http://ftpmirror.gnu.org/make/
List of required files:
* make-3.82.tar.bz2 (found at /home/kehoste/easybuild/sources/make-3.82.tar.bz2)
```

vs

```
[fetch_step method]
Download URLs (for files not available yet):
* http://ftpmirror.gnu.org/make/
List of required files:
* make-3.82.tar.bz2 (will be downloaded to /home/kehoste/easybuild/sources/make-3.82.tar.bz2)
```

Thoughts?

**Comment by pescobar** [comment]
@boegel I think printing the list of source urls would be fine. I would not go so far as doing a "fake http request" I don't think it's worth adding that extra complexity which can trigger errors which confuse the user. IMHO printing the list of source_urls is fine

Being picky, I would append the filename to the url so it prints something like (easier in case you want to copy& paste to try if the url works)

```
[fetch_step method]
Download URLs (for files not available yet):
* http://ftpmirror.gnu.org/make/make-3.82.tar.bz2
```

**Comment by boegel** [comment]
@pescobar: glueing the file names to the source URLs is going to be a problem if there are multiple URLs and multiple source files...

**Comment by pescobar** [comment]
@boegel I didn't realize about that. Then just ignore my suggestion I think it's fine as you suggest.

**Comment by verdurin** [comment]
@boegel - understood about this extra wrinkle. Your proposal to deal with it makes sense to me.

Just as a comparison, with RPMs you can use spectool to do things like verifying source URLs. Perhaps something like that would be useful for EB, which would be separate from this sort of dry-run test.

**Comment by boegel** [comment]
@pforai, @fgeorgatos, @verdurin, @gppezzi, @pescobar: updated the PR and description to reflect the current state

I still need to handle a couple of issues w.r.t. extensions (environment is printed over and over again when extensions are being handled, which is not nice when doing a dry run for a non-bare Perl easyconfig...)

feedback, suggestions?

File: easybuild/framework/easyblock.py [hunk 1]
```diff
@@ -1239,15 +1272,27 @@ def fetch_step(self, skip_checksums=False):
                 raise EasyBuildError("EasyBuild-version %s is newer than the currently running one. Aborting!",
                                      easybuild_version)
 
+        if build_option('extended_dry_run'):
+
+            print_msg("Available download URLs for missing sources/patches:", silent=self.silent, prefix=False)
```
Comment (verdurin) [inline easybuild/framework/easyblock.py]: I would change this to:

"Available download URLs for required sources/patches:"

I think I know why you've phrased it as "missing", but it sounds odd.

Comment (boegel) [inline easybuild/framework/easyblock.py]: @verdurin: I see your point. I'll drop the 'required' part, and just make it "Available download URLs for sources/patches:"

**Comment by boegel** [comment]
For me, the framework side of this is covered now, I'm happy with how this works and what it spits out (I updated the example output in the description).
More work will be needed to maximise the usefulness of the output we get, by doing some minor tweaks in the easyblocks (cfr. https://github.com/hpcugent/easybuild-easyblocks/pull/696 and https://github.com/hpcugent/easybuild-easyblocks/pull/685).

I have documentation for this in the works, will try and finish it ASAP (that's a blocker for getting this merged in).

I think it's time that this gets a thorough review... @wpoely86: are you up for taking a look at this beast?

**Comment by boegel** [comment]
draft documentation is available at http://boegel-eb.readthedocs.org/en/extended_dry_run/Extended_dry_run.html (see https://github.com/hpcugent/easybuild/pull/159)

File: easybuild/framework/easyblock.py [hunk 2]
```diff
@@ -235,8 +243,12 @@ def _init_log(self):
...
 
         this_module = inspect.getmodule(self)
-        self.log.info("This is easyblock %s from module %s (%s)",
-                      self.__class__.__name__, this_module.__name__, this_module.__file__)
+        eb_class = self.__class__.__name__
+        eb_loc = this_module.__file__
+        self.log.info("This is easyblock %s from module %s (%s)", eb_class, this_module.__name__, eb_loc)
+
+        if self.dry_run:
+            dry_run_msg("*** DRY RUN using '%s' easyblock (from %s) ***\n" % (eb_class, eb_loc), silent=self.silent)
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: this is really nitpicking but: `DRY RUN using easyblock '%s' `

And why not print the module name in case of dry run?

File: easybuild/framework/easyblock.py [hunk 3]
```diff
@@ -1253,15 +1293,27 @@ def fetch_step(self, skip_checksums=False):
...
+
+            dry_run_msg("Available download URLs for sources/patches:", silent=self.silent)
+            if self.cfg['source_urls']:
+                for source_url in self.cfg['source_urls']:
+                    dry_run_msg("  * %s/$source" % source_url, silent=self.silent)
+            else:
+                dry_run_msg('(none)', silent=self.silent)
+
+            # actual list of sources is printed via _obtain_file_dry_run method
+            dry_run_msg("\nList of sources:", silent=self.silent)
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: so print this over there?

Comment (wpoely86) [inline easybuild/framework/easyblock.py]: never mind, I understand :wink:

File: easybuild/framework/easyblock.py [hunk 4]
```diff
@@ -1575,7 +1662,68 @@ def sanity_check_step(self, custom_paths=None, custom_commands=None, extension=F
...
+            else:
+                commands = []
+                self.log.info("Using specified sanity check commands: %s" % commands)
+
+        for i, command in enumerate(commands):
+            # set command to default. This allows for config files with
+            # non-tuple commands
+            if not isinstance(command, tuple):
+                self.log.debug("Setting sanity check command to default")
+                command = (None, None)
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: shouldn't be:
```python
command = (command, None)
```

File: easybuild/framework/easyblock.py [hunk 5]
```diff
@@ -1575,7 +1662,68 @@ def sanity_check_step(self, custom_paths=None, custom_commands=None, extension=F
...
+            dry_run_msg("  (none)", silent=self.silent)
+
+    def _sanity_check_step(self, custom_paths=None, custom_commands=None, extension=False):
+        """Real version of sanity_check_step method."""
+        paths, path_keys_and_check, commands = self._sanity_check_step_common(custom_paths, custom_commands)
+
+        # check sanity check paths
         for key, check_fn in path_keys_and_check.items():
+            if key == 'dirs':
+                typ = "(non-empty) directory"
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: why replace this here but not at the `dry_run` version?

File: easybuild/framework/easyblock.py [hunk 6]
```diff
@@ -1708,19 +1832,28 @@ def make_module_step(self, fake=False):
         txt += self.make_module_footer()
 
         mod_filepath = self.module_generator.get_module_filepath(fake=fake)
-        write_file(mod_filepath, txt)
 
-        self.log.info("Module file %s written: %s", mod_filepath, txt)
+        if self.dry_run:
+            if not fake:
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: I love the mixture of `dry_run` and `fake`. I can only imagine the look on people's faces when they read this :wink:

File: easybuild/framework/easyblock.py [hunk 7]
```diff
@@ -1696,7 +1820,7 @@ def cleanup_step(self):
 
     def make_module_step(self, fake=False):
         """
-        Generate a module file.
+        Generate module file
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: maybe add a comment about the purpose of `fake`?

File: easybuild/framework/easyblock.py [hunk 8]
```diff
@@ -1823,7 +1956,25 @@ def run_step(self, step, methods):
...
             self.log.info("Running method %s part of step %s" % ('_'.join(m.func_code.co_names), step))
-            m(self)
+
+            if self.dry_run:
+
+                dry_run_msg("[%s method]" % m(self).__name__, silent=self.silent)
+
+                # if an known possible error occurs, just report it and continue
+                try:
+                    m(self)()
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: why is the last `()` needed?

File: easybuild/framework/easyblock.py [hunk 9]
```diff
@@ -1823,7 +1956,25 @@ def run_step(self, step, methods):
...
+                    if build_option('extended_dry_run_ignore_errors'):
+                        dry_run_msg("!!! WARNING !!! ignoring error: %s" % err, silent=self.silent)
+                        self.ignored_errors = True
+                    else:
+                        raise
+
+                dry_run_msg('', silent=self.silent)
+
+            else:
+                m(self)()
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: maybe a bit less empty lines in this part?

File: easybuild/framework/easyblock.py [hunk 10]
```diff
@@ -1840,46 +1991,46 @@ def get_step(tag, descr, substeps, skippable, initial=True):
 
         # list of substeps for steps that are slightly different from 2nd iteration onwards
         ready_substeps = [
-            (False, lambda x: x.check_readiness_step()),
-            (True, lambda x: x.make_builddir()),
-            (True, lambda x: env.reset_changes()),
-            (True, lambda x: x.handle_iterate_opts()),
+            (False, lambda x: x.check_readiness_step),
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: okay, this answers my previous question. Time for a new one: why the change?

File: easybuild/framework/easyblock.py [hunk 11]
```diff
@@ -1978,6 +2150,10 @@ def build_and_install_one(ecdict, init_env):
 
     try:
         app_class = get_easyblock_class(easyblock, name=name)
+
+        if dry_run:
+            print_dry_run_warning('below', silent=silent)
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: please add a comment about the why here

File: easybuild/framework/easyconfig/easyconfig.py [hunk 12]
```diff
@@ -1027,9 +1043,12 @@ def robot_find_easyconfig(name, version):
             if os.path.isfile(easyconfig_path):
                 _log.debug("Found easyconfig file for name %s, version %s at %s" % (name, version, easyconfig_path))
                 _easyconfig_files_cache[key] = os.path.abspath(easyconfig_path)
-                return _easyconfig_files_cache[key]
+                res = _easyconfig_files_cache[key]
```
Comment (wpoely86) [inline easybuild/framework/easyconfig/easyconfig.py]: Why this change?

File: easybuild/framework/easyconfig/easyconfig.py [hunk 13]
```diff
@@ -1014,12 +1028,14 @@ def robot_find_easyconfig(name, version):
     if key in _easyconfig_files_cache:
         _log.debug("Obtained easyconfig path from cache for %s: %s" % (key, _easyconfig_files_cache[key]))
         return _easyconfig_files_cache[key]
+
     paths = build_option('robot_path')
-    if not paths:
-        raise EasyBuildError("No robot path specified, which is required when looking for easyconfigs (use --robot)")
```
Comment (wpoely86) [inline easybuild/framework/easyconfig/easyconfig.py]: why is this error gone?

File: easybuild/tools/environment.py [hunk 14]
```diff
@@ -93,6 +94,12 @@ def setvar(key, value):
     _changes[key] = value
     _log.info("Environment variable %s set to %s (%s)", key, value, oldval_info)
 
+    if verbose and build_option('extended_dry_run'):
```
Comment (wpoely86) [inline easybuild/tools/environment.py]: why no `verbose` when not in dry run?

File: easybuild/tools/filetools.py [hunk 15]
```diff
@@ -154,22 +157,28 @@ def read_file(path, log_error=True):
...
     """Write given contents to file at given path (overwrites current file contents!)."""
-    f = None
+
+    # early exit in 'dry run' mode
+    if not forced and build_option('extended_dry_run'):
+        dry_run_msg("file written: %s" % path, silent=build_option('silent'))
+        return
+
+    handle = None
     # note: we can't use try-except-finally, because Python 2.4 doesn't support it as a single block
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: but now we can!

File: easybuild/tools/filetools.py [hunk 16]
```diff
@@ -604,33 +613,41 @@ def apply_patch(patch_file, dest, fn=None, copy=False, level=None):
...
+        # skip checking of files in dry run mode
+        patch_filename = os.path.basename(patch_file)
+        dry_run_msg("* applying patch file %s" % patch_filename, silent=build_option('silent'))
+
+    elif not os.path.isfile(patch_file):
         raise EasyBuildError("Can't find patch %s: no such file", patch_file)
-        return
 
-    if fn and not os.path.isfile(fn):
+    elif fn and not os.path.isfile(fn):
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: sure this needs to be `elif`?  They all need to be checked?

File: easybuild/tools/filetools.py [hunk 17]
```diff
@@ -604,33 +613,41 @@ def apply_patch(patch_file, dest, fn=None, copy=False, level=None):
...
-        except IOError, err:
-            raise EasyBuildError("Failed to copy %s to dir %s: %s", patch_file, dest, err)
-            return
+        if build_option('extended_dry_run'):
+            dry_run_msg("  %s copied to %s" % (patch_file, dest), silent=build_option('silent'))
+        else:
+            try:
+                shutil.copy2(patch_file, dest)
+                _log.debug("Copied patch %s to dir %s" % (patch_file, dest))
+                return 'ok'
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: say what?

File: easybuild/tools/filetools.py [hunk 18]
```diff
@@ -604,33 +613,41 @@ def apply_patch(patch_file, dest, fn=None, copy=False, level=None):
...
+            except IOError, err:
+                raise EasyBuildError("Failed to copy %s to dir %s: %s", patch_file, dest, err)
 
     # use absolute paths
     apatch = os.path.abspath(patch_file)
     adest = os.path.abspath(dest)
 
-    if not level:
+    patch_level = level
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: why is this?

File: easybuild/tools/filetools.py [hunk 19]
```diff
@@ -604,33 +613,41 @@ def apply_patch(patch_file, dest, fn=None, copy=False, level=None):
...
+    if patch_level is None and build_option('extended_dry_run'):
+        patch_level = '<derived>'
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: shouldn't we simply skip the actual calling of `patch`?

File: easybuild/tools/filetools.py [hunk 20]
```diff
@@ -658,17 +674,32 @@ def apply_patch(patch_file, dest, fn=None, copy=False, level=None):
...
     result = run.run_cmd(patch_cmd, simple=True)
     if not result:
         raise EasyBuildError("Patching with patch %s failed", patch_file)
-        return
 
     return result
 
 
+def apply_regex_substitutions(path, regex_subs):
+    """Apply specified list of regex substitutions."""
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: document the arguments

File: easybuild/tools/filetools.py [hunk 21]
```diff
@@ -658,17 +674,32 @@ def apply_patch(patch_file, dest, fn=None, copy=False, level=None):
...
+
+    # early exit in 'dry run' mode
+    if build_option('extended_dry_run'):
+        dry_run_msg("patching file %s" % path, silent=build_option('silent'))
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: maybe change to `applying regex...` to clearly differentiate between patching from the easyconfig and the easyblock.

File: easybuild/tools/filetools.py [hunk 22]
```diff
@@ -658,17 +674,32 @@ def apply_patch(patch_file, dest, fn=None, copy=False, level=None):
...
+        return
+
+    for i, (regex, subtxt) in enumerate(regex_subs):
+        regex_subs[i] = (re.compile(regex), subtxt)
```
Comment (wpoely86) [inline easybuild/tools/filetools.py]: what about specifying `re.M` or such?

File: easybuild/tools/run.py [hunk 23]
```diff
@@ -73,6 +74,22 @@ def run_cmd(cmd, log_ok=True, log_all=False, simple=False, inp=None, regexp=True
     - path is the path run_cmd should chdir to before doing anything
```
Comment (wpoely86) [inline easybuild/tools/run.py]: all options should be documented

File: easybuild/tools/config.py [hunk 24]
```diff
@@ -517,46 +521,3 @@ def module_classes():
 def read_environment(env_vars, strict=False):
     """NO LONGER SUPPORTED: use read_environment from easybuild.tools.environment instead"""
     _log.nosupport("read_environment has moved to easybuild.tools.environment", '2.0')
-
-
-def set_tmpdir(tmpdir=None, raise_error=False):
```
Comment (wpoely86) [inline easybuild/tools/config.py]: shouldn't we do a deprecated warning thingie? As good policy?

File: easybuild/tools/toolchain/toolchain.py [hunk 25]
```diff
@@ -363,79 +378,133 @@ def is_dep_in_toolchain_module(self, name):
         """Check whether a specific software name is listed as a dependency in the module for this toolchain."""
         return any(map(lambda m: self.mns.is_short_modname_for(m, name), self.toolchain_dep_mods))
 
-    def _prepare_dependency_external_module(self, dep):
+    def _simulated_load_dependency_module(self, name, version, metadata):
         """Set environment variables picked up by utility functions for dependencies specified as external modules."""
```
Comment (wpoely86) [inline easybuild/tools/toolchain/toolchain.py]: document the args

File: easybuild/framework/easyblock.py (hunk 2 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: your rephrasing suggestion makes less sense to me if I read it out loud than what I have now...

I'll print the module name as well, you're right

File: easybuild/framework/easyblock.py (hunk 4 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: no, the 1st `None` is important, it basically results in using `self.name.lower()` as default command (see below)

File: easybuild/framework/easyblock.py (hunk 5 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: to avoid duplicate code? ;)

will check, this could make the output of the dry run sanity_check_step clearer, indeed

File: easybuild/framework/easyblock.py (hunk 8 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: the method as passed by 'name', so `()` is needed to actually *call* the method, rather than just referring to it, I'll add a comment to clarify that

File: easybuild/framework/easyblock.py (hunk 10 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: by returning the method itself, I can access `.func_code.co_names`

File: easybuild/framework/easyconfig/easyconfig.py (hunk 13 above)
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: serves no use, `robot_path` is always defined via `main()`

File: easybuild/framework/easyconfig/easyconfig.py (hunk 12 above)
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: mostly style, returning somewhere 'deep' in the function results in code that's harder to read or understand at a glance (it's not immediately clear if there are other `return`s for example)

File: easybuild/tools/environment.py (hunk 14 above)
Comment (boegel) [inline easybuild/tools/environment.py]: the `verbose` indicates whether or not defining this environment variables should be printed under `-x` or not

in some places, defining the environment variables is considered 'internal', i.e. it serves a purposes but is mostly irrelevant so shouldn't 'pollute' the dry run output

example: all the `$EBVAR*` environment variables (which we should get rid of), there are others too

Comment (boegel) [inline easybuild/tools/environment.py]: I'll clarify this by updating the `setvar` docstring

File: easybuild/tools/filetools.py (hunk 15 above)
Comment (boegel) [inline easybuild/tools/filetools.py]: well, using `with` (context manager) avoids the need for `try-except-finally` all together, so let me use that instead :)

File: easybuild/tools/filetools.py (hunk 16 above)
Comment (boegel) [inline easybuild/tools/filetools.py]: well, it's just cleaner; if the `elif` above passes, we'll `raise` anyway

File: easybuild/tools/filetools.py (hunk 17 above)
Comment (boegel) [inline easybuild/tools/filetools.py]: hehe, should be `return True`, will fix

File: easybuild/tools/filetools.py (hunk 18 above)
Comment (boegel) [inline easybuild/tools/filetools.py]: there was a `patch_level = level` deeper down, so I moved it up

it is a bit weird though, maybe sticking to `level` is clearer indeed, will fix

File: easybuild/tools/filetools.py (hunk 19 above)
Comment (boegel) [inline easybuild/tools/filetools.py]: that's skipped already (`run_cmd` is modified to not execute any commands unless forced)

trying to derive the patch level should never be tried though, since that requires the patch file to be available, and to find/read it, etc.

File: easybuild/tools/filetools.py (hunk 22 above)
Comment (boegel) [inline easybuild/tools/filetools.py]: this function currently only supports patching by lines, so no multiline patching support yet (it's rather rare to require that at runtime)

we can enhance this function later to support that

File: easybuild/tools/config.py (hunk 24 above)
Comment (boegel) [inline easybuild/tools/config.py]: this is internal to framework, if someone is using this in an easyblock, they're doing something weird or something they shouldn't be doing ;)

File: easybuild/framework/easyblock.py (hunk 4 above)
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: I was more thing in the case of, what if command is just a string?

File: easybuild/tools/toolchain/toolchain.py [hunk 26]
```diff
@@ -363,79 +378,139 @@ def is_dep_in_toolchain_module(self, name):
...
+            if tcmods_exists:
+                self.modules_tool.load([tc_mod], silent=silent)
+                dry_run_msg("module load %s" % tc_mod, silent=silent)
+            else:
+                dry_run_msg("module load %s [SIMULATED]" % tc_mod, silent=silent)
+                # provide fake prefix so $EBROOT* gets defined
+                fake_prefix = os.path.join(fake_root, tc_mod)
+                self._simulated_load_dependency_module(self.name, self.version, {'prefix': fake_prefix})
+
+                # also simulate loads for toolchain dependencies, if required information is available
```
Comment (wpoely86) [inline easybuild/tools/toolchain/toolchain.py]: shouldn't simulated the load of the toolchain deps before loading the toolchain itself?

File: easybuild/tools/toolchain/toolchain.py [hunk 27]
```diff
@@ -363,79 +378,139 @@ def is_dep_in_toolchain_module(self, name):
...
+                    # provide fake prefix so $EBROOT* gets defined
+                    fake_prefix = os.path.join(fake_root, mod_name)
+                    self._simulated_load_dependency_module(dep['name'], dep['version'], {'prefix': fake_prefix})
+        else:
+            # load modules for all dependencies
+            dep_mods = [dep['short_mod_name'] for dep in self.dependencies]
+            self.log.debug("Loading modules for dependencies: %s" % dep_mods)
+            self.modules_tool.load(dep_mods, silent=silent)
 
         # define $EBROOT* and $EBVERSION* for external modules, if metadata is available
```
Comment (wpoely86) [inline easybuild/tools/toolchain/toolchain.py]: this comment is no longer accurate, they are already loaded?

Comment (wpoely86) [inline easybuild/tools/toolchain/toolchain.py]: a no, just a strange diff here.

File: test/framework/options.py [hunk 28]
```diff
@@ -1898,6 +1904,75 @@ def test_review_pr(self):
...
+        self.mock_stdout(False)
+        self.assertEqual(len(stdout), 0)
+
+        msg_regexs = [
+            re.compile(r"the actual build \& install procedure that will be performed may diverge", re.M),
+            re.compile(r"^\*\*\* DRY RUN using 'EB_toy' easyblock", re.M),
+            re.compile(r"^== COMPLETED: Installation ended successfully", re.M),
+        ]
+        ignoring_error_regex = re.compile(r"!!! WARNING !!! ignoring error", re.M)
+        ignored_error_regex = re.compile(r"!!! WARNING !!! One or more errors were ignored, see warnings above", re.M)
```
Comment (boegel) [inline test/framework/options.py]: hmm, this is broken

File: easybuild/framework/easyblock.py [hunk 29]
```diff
@@ -246,6 +260,20 @@ def close_log(self):
...
 
     #
+    # INIT UTILITIES
+    #
+    def init_dry_run(self):
+        """Initialise easyblock instance for performing a dry run."""
+
+        # replace build/install dir
+        self.builddir = os.path.join(tempfile.gettempdir(), '__ROOT__', self.builddir.lstrip(os.path.sep))
+        dry_run_msg("Using fake build directory: %s" % self.builddir, silent=self.silent)
```
Comment (wpoely86) [inline easybuild/framework/easyblock.py]: We should display this more clearly: keep using a 'fake' build and install dir but displaying them nicely like: `$EASYBUILD_BUILDDIR/...`.

Comment (wpoely86) [inline easybuild/framework/easyblock.py]: Like @rjeschmi asked

File: easybuild/tools/toolchain/toolchain.py [hunk 30]
```diff
@@ -363,79 +378,139 @@ def is_dep_in_toolchain_module(self, name):
...
 
-    def _prepare_dependencies(self):
+    def _load_toolchain_module(self, silent=False):
+        """Load toolchain module."""
+
+        tc_mod = self.det_short_module_name()
+
+        if self.dry_run:
+
+            fake_root = os.path.join(tempfile.gettempdir(), '__fake__')
```
Comment (wpoely86) [inline easybuild/tools/toolchain/toolchain.py]: Using both `__root__` and `__fake__` is confusing. Let's show the `$EASYBUILD_INSTALLPATH_MODULES`, as @tbrown-curc asked

File: easybuild/framework/easyblock.py (hunk 4 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: Huh, ok, that makes a lot of sense, good suggestion. :-)

File: easybuild/tools/toolchain/toolchain.py (hunk 26 above)
Comment (boegel) [inline easybuild/tools/toolchain/toolchain.py]: yep, that's better, will change

File: easybuild/tools/toolchain/toolchain.py (hunk 30 above)
Comment (boegel) [inline easybuild/tools/toolchain/toolchain.py]: I think the idea was to show `$EBROOTFOO` instead.

File: easybuild/framework/easyblock.py (hunk 29 above)
Comment (boegel) [inline easybuild/framework/easyblock.py]: In every dry run message being printed, the fake build/install dir is now replaced with the original build/install dir, so no more `__ROOT__` in the output, but the build/install dir is still replaced as was the case here to make sure only paths in the tmpdir are used/created/touched

**Comment by boegel** [comment]
@wpoely86, @rjeschmi, @tbrown-curc: remarks w.r.t. `__ROOT__` and `__fake__` paths used in output fixed, updated the example in the PR description and at http://boegel-eb.readthedocs.org/en/extended_dry_run/Extended_dry_run_examples.html, looks a lot better now!

**Comment by boegel** [comment]
I gave this another full review myself, and added some final minor tweaks.

Since the documentation is also update (see https://github.com/hpcugent/easybuild/pull/159), I feel this is ready to merge as soon as Jenkins gives the green light.

Thanks to everyone who contributed to this in one way or another, in my view this is going to be a killer feature.
